google/cloud/bigtable/internal/strong_type.h

```diff
@@ -56,8 +56,7 @@ namespace internal {
 template <typename T, typename Parameter>
 class StrongType {
  public:
-  explicit StrongType(T const& value) : value_(value) {}
-  explicit StrongType(T&& value) : value_(std::move(value)) {}
+  explicit StrongType(T value) : value_(std::move(value)) {}
```

Comment:
Overall, I like the rest of the changes, but is this an improvement? According to this [answer](https://stackoverflow.com/a/51706522) you're going from one copy or move to copy/move or move/move for lvalue and rvalues respectively.

Reply:
Thanks.

I think the improvement is in simplicity (one simple function vs two), and consistency (in almost all of the functions in this code base, arguments are passed by value and moved from). So this class follows that pattern.

While I agree with you that there's potentially one extra move in this version, the simpler and more consistent version is worth it, IMO. 

But I also don't feel very strongly about this particular situation. I'm happy to revert it if others prefer the first version.

Reply:
Makes sense. In that case, could we still use perfect forwarding here as shown in the link? While perfect forwarding [doesn't apply to class template parameters](https://timsong-cpp.github.io/cppwp/temp.deduct.call#def:forwarding_reference) they should still apply to constructors.

Reply:
Perfect forwarding would work fine if we make the constructor a template. I'm happy to make that change, but I'm not sure it's worth it. Are you concerned about the potentially unnecessary move?

FWIW, here's the diff of what it would look like:

```diff
diff --git a/google/cloud/bigtable/internal/strong_type.h b/google/cloud/bigtable/internal/strong_type.h
index 7d36db7f..c0ad8d49 100644
--- a/google/cloud/bigtable/internal/strong_type.h
+++ b/google/cloud/bigtable/internal/strong_type.h
@@ -16,6 +16,8 @@
 #define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_STRONG_TYPE_H_
 
 #include "google/cloud/bigtable/version.h"
+#include <type_traits>
+#include <utility>
 
 namespace google {
 namespace cloud {
@@ -56,7 +58,10 @@ namespace internal {
 template <typename T, typename Parameter>
 class StrongType {
  public:
-  explicit StrongType(T value) : value_(std::move(value)) {}
+  template <typename U,
+            typename std::enable_if<std::is_constructible<T, U&&>::value,
+                                    int>::type = 0>
+  explicit StrongType(U&& value) : value_(std::forward<U>(value)) {}
   T& get() { return value_; }
   T const& get() const { return value_; }
 
```

plus tests. If we decide to go this way, I think I'd prefer to do it in a separate PR that would be more focused making the change and adding the unit test.

Reply:
Agreed about the change being in a separate PR, just brought it up for discussion. I doubt there will be a performance hit with what you have.

Reply:
@remyabel so where did we land on this one? I'm happy to leave this constructor taking a `T` by value as in this PR. We can always change it to a templated constructor with a forwarding reference if there's ever a performance issue caused by the one potential extra move. Sound good? Other thoughts?

Thanks for raising the discussion, btw.

Reply:
Yes I'm fine with leaving it as is. I did a [quick-bench](http://quick-bench.com/b7W7B2tyEYvnPSMLs882iqNiyzI) and while there is a difference it's negligible.